### PR TITLE
Merkle tree implementation

### PR DESCRIPTION
**Pull request by terencechain**
Building off #3571, this PR implements merkle tree + test

Reference: https://github.com/ethereum/eth2.0-specs/blob/50209ea806820f83ceff3f21e1578191db0a71e5/specs/light_client/merkle_proofs.md#generalized-merkle-tree-index

### CODE REVIEW

File: shared/trieutil/merkle_trie.go [hunk 1]
```diff
@@ -35,3 +40,38 @@ func PrevPowerOf2(n int) int {
...
+
+	for i := 0; i < len(parents); i++ {
+		parents[i] = params.BeaconConfig().ZeroHash[:]
+	}
+	for i := 0; i < len(paddedLeaves); i++ {
+		paddedLeaves[i] = params.BeaconConfig().ZeroHash[:]
+	}
+
+	merkleTree := make([][]byte, 0, len(parents)+len(leaves)+len(paddedLeaves))
+	merkleTree = append(merkleTree, parents...)
```
Comment (rauljordan) [inline shared/trieutil/merkle_trie.go]: I wonder if copy is faster than append here? Given the lengths are all known

Comment (terencechain) [reply shared/trieutil/merkle_trie.go]: Will benchmark :)

Comment (terencechain) [reply shared/trieutil/merkle_trie.go]: Copy is faster. Tested with 2**20 leaves.
**Append:**
BenchmarkMerkleTree_Generate-4   	      10	 226862027 ns/op
**Copy:**
BenchmarkMerkleTree_Generate-4   	      20	  67944152 ns/op

Comment (terencechain) [reply shared/trieutil/merkle_trie.go]: Updated to use copy, thanks for the great suggestion!

**Pull request merged**